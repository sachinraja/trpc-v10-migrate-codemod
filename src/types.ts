import { ImportDeclarationStructure, OptionalKind } from 'ts-morph'

export interface MigrateConfig {
	/**
	 * namespace of the tRPC React hooks
	 * @default ['trpc']
	 */
	reactNamespace: string[]
	/**
	 * namespace of the tRPC callers
	 * @default ['caller']
	 */
	callerNamespace: string[]

	/**
	 * the function you use to create your routers
	 * @default ['router']
	 */
	routerFactory: string[]
	/**
	 * @default 'tsconfig.json'
	 */
	tsconfigPath: string
	/**
	 * @default 'package.json'
	 */
	packageJSONPath: string
	/**
	 * variable name of the base procedure to chain new ones from
	 * @default 't.procedure'
	 */
	baseProcedure: string
	/**
	 * imports to add to the top of every file with a transformed router
	 * @default []
	 */
	serverImports: OptionalKind<ImportDeclarationStructure>[]
	/**
	 * imports to remove from the top of every file with a transformed router
	 */
	removeServerImports: {
		moduleSpecifier: string
		namedImports: string[]
	}[]
}
