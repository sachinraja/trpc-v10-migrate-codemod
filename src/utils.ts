import { exec } from 'node:child_process'
import { existsSync } from 'node:fs'
import fs from 'node:fs/promises'
import path from 'node:path'
import { promisify } from 'node:util'
import { CodeBlockWriter, Node, ObjectLiteralElementLike, SyntaxKind } from 'ts-morph'

export const execa = promisify(exec)

export const getStringFromStringOrArrayLiteral = (node: Node) => {
	if (Node.isStringLiteral(node)) {
		return node.getLiteralValue()
	}
	if (Node.isArrayLiteralExpression(node)) {
		return node
			.getElements()[0]
			.asKindOrThrow(SyntaxKind.StringLiteral)
			.getLiteralText()
	}
}

export const writeValueFromObjectLiteralElement = (
	writer: CodeBlockWriter,
	node: ObjectLiteralElementLike,
) => {
	if (Node.isMethodDeclaration(node)) {
		if (node.isAsync()) {
			writer.write('async ')
		}
		const parameters = node
			.getParameters()
			.map((parameter) => parameter.getText())
		return writer
			.write('(')
			.write(parameters.join(', '))
			.write(') => ')
			.inlineBlock(() => {
				writer.write(node.getBodyText() ?? '')
			})
	}
	if (Node.isPropertyAssignment(node)) {
		return writer.write(node.getChildAtIndex(2).getText())
	}
	if (Node.isShorthandPropertyAssignment(node)) {
		return writer.write(node.getName())
	}
}

export const getStringLiteralOrText = (node: Node) => {
	if (Node.isStringLiteral(node)) {
		return node.getLiteralValue()
	}
	return node.getText()
}

export const getStringHash = (string_: string) => {
	let hash = 0
	for (let index = 0, length = string_.length; index < length; index++) {
		// eslint-disable-next-line unicorn/prefer-code-point
		const chr = string_.charCodeAt(index)
		hash = (hash << 5) - hash + chr
		hash = Math.trunc(hash)
	}
	return hash >>> 0
}

export const getDefinedProperties = <TObject extends Record<string, unknown>>(
	object: TObject,
): TObject =>
	Object.fromEntries(
		Object.entries(object).filter(([, v]) => v !== undefined),
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
	) as any

export const modifyVersions = async (packageJSON: string) => {
	const directory = process.cwd()
	if (existsSync(path.join(directory, packageJSON))) {
		const pkgs: {
			dependencies?: Record<string, string>
			devDependencies?: Record<string, string>
		} = JSON.parse(
			await fs.readFile(path.join(directory, packageJSON), 'utf8'),
		)
		let shouldPrint = false
		shouldPrint = !!(
			pkgs.dependencies
			&& (await updateOutdatedPackages(directory, pkgs.dependencies))
		)
		shouldPrint = !!(
			pkgs.devDependencies
			&& (await updateOutdatedPackages(directory, pkgs.devDependencies, true))
		)
		!shouldPrint && console.log('Packages are up to date!')
	}
}

const updateOutdatedPackages = async (
	directory: string,
	deps: Record<string, string>,
	isDevelopment?: boolean,
): Promise<boolean> => {
	const results = Object.entries(deps)
		.filter(
			([packageName, packageVersion]) =>
				(packageName.startsWith('@trpc/')
					&& !packageVersion.includes('proxy'))
				|| (packageName === 'solid-trpc' && !packageVersion.includes('beta')),
		)
		.map(([packageName]) => `${packageName}@next`)
	if (results.length > 0) {
		console.log(
			`Updating ${isDevelopment ? 'dev' : ''}Dependencies: ${
				results.join(
					', ',
				)
			}`,
		)
		await execa(
			`npm install ${results.join(' ')}${isDevelopment ? ' -D' : ''}`,
			{
				cwd: directory,
			},
		)
	}
	return results.length > 0
}
