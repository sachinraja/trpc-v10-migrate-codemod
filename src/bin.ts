#!/usr/bin/env node
import { Command, Option, runExit } from 'clipanion'
import { transformv10Migration } from './index.js'
import { getDefinedProperties, importsMappingToImportDeclaration } from './utils.js'

runExit(
	class extends Command {
		tsconfigPath = Option.String('--tsconfig-path', { description: 'filepath of tsconfig.json' })

		reactNamespace = Option.Array('--react-namespace', { description: 'namespace of your tRPC React hooks' })
		callerNamespace = Option.Array('--caller-namespace', {
			description: 'namespace of your tRPC createCallers',
		})
		contextNamespace = Option.Array('--context-namespace', {
			description: 'namespace of your tRPC context helpers',
		})

		routerFactory = Option.Array('--router-factory', {
			description: 'name of the function used to create your routers',
		})
		baseProcedure = Option.String('--base-procedure', {
			description: 'the variable name of the base procedure to chain new ones from',
		})
		imports = Option.Array('--import', [], {
			description: 'imports to add to the top of every file with a transformed router',
		})
		removeImports = Option.Array('--remove-import', [], {
			description: 'imports to remove from the top of every file with a transformed router',
		})

		static usage = Command.Usage({
			description: 'migrate your codebase from tRPC v9 to v10',
			examples: [
				[
					'basic example',
					'$0',
				],
				[
					'with configuration (the defaults are shown)',
					'$0 --tsconfig-path tsconfig.json --trpc-namespace trpc --router-factory router --base-procedure t.procedure',
				],
				['with injected server imports', '$0 --import t:~/server/trpc --import adminProcedure:~/server/trpc'],
				['with removed imports', '$0 --remove-import createRouter:~/server/trpc'],
			],
		})

		async execute() {
			console.log('migrating...')

			await transformv10Migration(getDefinedProperties({
				reactNamespace: this.reactNamespace,
				callerNamespace: this.callerNamespace,
				contextNamespace: this.contextNamespace,
				routerFactory: this.routerFactory,
				tsconfigPath: this.tsconfigPath,
				baseProcedure: this.baseProcedure,
				serverImports: importsMappingToImportDeclaration(
					this.imports,
				),
				removeServerImports: importsMappingToImportDeclaration(
					this.removeImports,
				),
			}))
		}
	},
)
