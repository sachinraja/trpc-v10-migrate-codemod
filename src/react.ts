import { CallExpression, Node, ObjectLiteralElementLike, SyntaxKind } from 'ts-morph'
import { normalizeProcedurePath } from './utils.js'

export const contextHelpers = [
	'invalidateQueries',
	'cancelQuery',
	'fetchQuery',
	'fetchInfiniteQuery',
	'getQueryData',
	'getInfiniteQueryData',
	'prefetchQuery',
	'prefetchInfiniteQuery',
	'refetchQueries',
	'setQueryData',
	'setInfiniteQueryData',
] as const

export type OldContextHelper = typeof contextHelpers[number]
export const contextHelpersToRename: Record<OldContextHelper, string> = {
	invalidateQueries: 'invalidate',
	cancelQuery: 'cancel',
	fetchQuery: 'fetch',
	fetchInfiniteQuery: 'fetchInfinite',
	getQueryData: 'getData',
	getInfiniteQueryData: 'getInfiniteData',
	prefetchQuery: 'prefetch',
	prefetchInfiniteQuery: 'prefetchInfinite',
	refetchQueries: 'refetch',
	setQueryData: 'setData',
	setInfiniteQueryData: 'setInfiniteData',
}

export const handleContextHelperCall = (callExpression: CallExpression) => {
	const arguments_ = callExpression.getArguments()
	const [pathArgument] = arguments_

	let rawPath: string | undefined
	let requiresUndefinedInput = true
	if (Node.isStringLiteral(pathArgument)) {
		rawPath = pathArgument.getLiteralValue()
		callExpression.removeArgument(pathArgument)
		if (arguments_.length === 1) requiresUndefinedInput = false
	} else if (Node.isArrayLiteralExpression(pathArgument)) {
		const elements = pathArgument.getElements()

		const pathElement = elements[0]
		if (!Node.isStringLiteral(pathElement)) return { requiresUndefinedInput: false }
		rawPath = pathElement.getLiteralText()
		callExpression.insertArguments(0, elements.slice(1).map(it => it.getText()))

		if (elements.length > 1 || arguments_.length === 1) {
			requiresUndefinedInput = false
		}
		callExpression.removeArgument(pathArgument)
	}

	const path = rawPath ? normalizeProcedurePath(rawPath) : undefined
	return { path, requiresUndefinedInput }
}

export const handleReactHookCall = (type: string, callExpression: CallExpression) => {
	if (type === 'useQuery') {
		const arguments_ = callExpression.getArguments()
		const [, configArgument] = arguments_

		if (Node.isObjectLiteralExpression(configArgument)) {
			const trpcPropertyNames = ['context', 'ssr']

			const trpcProperties = trpcPropertyNames
				.map((property) => configArgument.getProperty(property))
				.filter((node): node is ObjectLiteralElementLike => !!node)

			if (trpcProperties.length > 0) {
				const trpcProperty = configArgument.getProperty('trpc')?.asKind(SyntaxKind.ObjectLiteralExpression)
					?? configArgument.addPropertyAssignment({ name: 'trpc', initializer: '{}' }).getInitializerIfKind(
						SyntaxKind.ObjectLiteralExpression,
					)

				if (trpcProperty) {
					trpcProperty.addProperties(trpcProperties.map((property) => property.getText()))
				}
			}

			for (const property of trpcProperties) {
				property.remove()
			}

			configArgument.formatText()
		}

		const { path, requiresUndefinedInput } = handleContextHelperCall(callExpression)
		if (requiresUndefinedInput) {
			callExpression.insertArgument(0, 'undefined')
		}
		return path
	}

	if (type === 'useMutation') {
		// no input, so no need to handle undefined input
		const { path } = handleContextHelperCall(callExpression)
		return path
	}
}
