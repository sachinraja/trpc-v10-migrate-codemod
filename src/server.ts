import {
	CallExpression,
	CodeBlockWriter,
	Node,
	Project,
	SourceFile,
	SyntaxKind,
	VariableDeclarationKind,
} from 'ts-morph'
import { MigrateConfig } from './types.js'
import { getStringHash, getStringLiteralOrText, writeValueFromObjectLiteralElement } from './utils.js'

interface ProcedureUnit {
	tag: 'procedure'
	type: string
	pathText: string
	options?: Node
	middlewares: MiddlewareUnit[]
	middlewaresHash?: string
}

interface RouterUnit {
	tag: 'router'
	prefix?: string
	identifier: string
}

interface MiddlewareUnit {
	tag: 'middleware'
	id: string
	hash: number
	body: string
}

type Unit = ProcedureUnit | RouterUnit | MiddlewareUnit

const handleRouterPropertyAccessor = (
	options: { node: Node; arguments: Node[]; middlewares: MiddlewareUnit[] },
): Unit => {
	const { node, arguments: arguments_, middlewares } = options

	const propertyAccessorText = node.getText()
	if (propertyAccessorText === 'merge') {
		if (arguments_.length === 1) {
			return {
				tag: 'router',
				identifier: arguments_[0].getText(),
			}
		}

		const [prefix, router] = arguments_
		return {
			tag: 'router',
			prefix: getStringLiteralOrText(prefix),
			identifier: router.getText(),
		}
	}

	if (propertyAccessorText === 'middleware') {
		const [function_] = arguments_
		const middlewareBody = function_.getText()
		const hash = getStringHash(middlewareBody)
		return {
			tag: 'middleware',
			id: `middleware_${hash}`,
			hash,
			body: middlewareBody,
		}
	}

	const [path, optionsNode] = arguments_
	return {
		tag: 'procedure',
		type: node.getText(),
		pathText: getStringLiteralOrText(path),
		options: optionsNode,
		middlewares,
		middlewaresHash: middlewares.length > 0 ? middlewares.map((v) => v.id).join(',') : undefined,
	}
}

interface GetRouterProceduresOptions {
	node: CallExpression
	units?: Unit[]
	middlewares?: MiddlewareUnit[]
}

export const getRouterUnits = (
	options: GetRouterProceduresOptions,
): { units: Unit[]; topNode: Node } => {
	const { node, units = [], middlewares = [] } = options

	const propertyAccessParent = node.getParentIfKind(SyntaxKind.PropertyAccessExpression)

	if (!propertyAccessParent) return { units, topNode: node }
	const propertyAccessor = propertyAccessParent.getChildAtIndex(2)

	const callExpressionParent = propertyAccessParent.getParentIfKind(SyntaxKind.CallExpression)
	if (!callExpressionParent) return { units, topNode: propertyAccessParent }

	const unit = handleRouterPropertyAccessor({
		node: propertyAccessor,
		arguments: callExpressionParent.getArguments(),
		middlewares,
	})
	units.push(unit)

	const newMiddlewares = unit.tag === 'middleware' ? [...middlewares, unit] : middlewares

	return getRouterUnits({ node: callExpressionParent, units, middlewares: newMiddlewares })
}

type ProcedureOrRouterRecord = Record<string, { procedure: ProcedureUnit } | { router: RouterShape }>
interface RouterShape extends Pick<RouterUnit, 'tag' | 'prefix'> {
	units: ProcedureOrRouterRecord
	text?: string
}

type MiddlewareProcedureIdMap = Map<MiddlewareUnit[], string>

const addProcedure = (shape: RouterShape, procedureUnit: ProcedureUnit, pathParts: string[], index = 0) => {
	if (pathParts.length - 1 === index) {
		shape.units[pathParts[index]] = {
			...shape.units[pathParts[index]],
			procedure: procedureUnit,
		}
		return
	}

	let nextRouter: RouterShape
	if (!('router' in (shape.units[pathParts[index]] ?? {}))) {
		nextRouter = {
			tag: 'router',
			units: {},
		}
		shape.units[pathParts[index]] = {
			...shape.units[pathParts[index]],
			router: nextRouter,
		}
	} else {
		nextRouter = (shape.units[pathParts[index]] as { router: RouterShape }).router
	}

	addProcedure(nextRouter, procedureUnit, pathParts, index + 1)
}

const addRouter = (shape: RouterShape, routerUnit: RouterUnit, pathParts: string[], index = 0) => {
	if (pathParts.length - 1 === index) {
		shape.units[pathParts[index]] = {
			...shape.units[pathParts[index]],
			router: {
				tag: 'router',
				prefix: pathParts[index],
				text: routerUnit.identifier,
				units: {},
			},
		}
		return
	}

	let nextRouter: RouterShape
	if (!('router' in (shape.units[pathParts[index]] ?? {}))) {
		nextRouter = {
			tag: 'router',
			prefix: pathParts[index],
			units: {},
		}
		shape.units[pathParts[index]] = {
			...shape.units[pathParts[index]],
			router: nextRouter,
		}
	} else {
		nextRouter = (shape.units[pathParts[index]] as { router: RouterShape }).router
	}

	addRouter(nextRouter, routerUnit, pathParts, index + 1)
}

const writeProcedure = (options: { writer: CodeBlockWriter; unit: ProcedureUnit; baseProcedureId: string }) => {
	const { writer, unit, baseProcedureId } = options
	const { type, options: optionsNode } = unit

	writer.write(baseProcedureId)
	if (Node.isObjectLiteralExpression(optionsNode)) {
		for (const procedureOption of ['input', 'output', 'meta']) {
			const property = optionsNode?.getProperty(procedureOption)
			if (!property) continue

			writer.write(`.${procedureOption}(`)
			writeValueFromObjectLiteralElement(writer, property)
			writer.write(`)`)
		}

		const resolver = optionsNode?.getProperty('resolve')
		if (resolver) {
			writer.write(`.${type}(`)
			writeValueFromObjectLiteralElement(writer, resolver)
			writer.write(')')
		}
	} else {
		writer.write(`.${type}(${optionsNode?.getText() ?? ''})`)
	}
	return
}

const writeShape = (
	options: {
		writer: CodeBlockWriter
		procedureOrRouter: { procedure: ProcedureUnit } | { router: RouterShape }
		path?: string
		middlewaresProcedureIdMap: MiddlewareProcedureIdMap
		config: MigrateConfig
	},
) => {
	const { writer, procedureOrRouter, path, middlewaresProcedureIdMap, config } = options

	if ('procedure' in procedureOrRouter) {
		if (path) {
			writer.write(path).write(': ')
		}
		writeProcedure({
			writer,
			unit: procedureOrRouter.procedure,
			baseProcedureId: middlewaresProcedureIdMap.get(procedureOrRouter.procedure.middlewares) ?? config.baseProcedure,
		})

		if (path) {
			writer.write(',')
		}
	}

	if ('router' in procedureOrRouter) {
		if (path) {
			writer.write(path).write(': ')
		}

		if ('text' in procedureOrRouter) {
			const { text } = procedureOrRouter
			writer.write(`${text},`)
			return
		}

		writer.write(`t.router(`).inlineBlock(() => {
			for (const [path, nestedShape] of Object.entries(procedureOrRouter.router.units)) {
				writeShape({
					writer,
					procedureOrRouter: nestedShape,
					path,
					middlewaresProcedureIdMap,
					config,
				})
			}
		}).write(')')

		if (path) {
			writer.write(',')
		}
	}

	writer.newLine()
}

export const writeNewRouter = (
	options: {
		units: Unit[]
		sourceFile: SourceFile
		topNode: Node
		config: MigrateConfig
		project: Project
	},
) => {
	const { units, sourceFile, topNode, config, project } = options

	const procedureUnits = units.filter((unit): unit is ProcedureUnit => unit.tag === 'procedure')
	const routerUnits = units.filter((unit): unit is RouterUnit => unit.tag === 'router')

	const procedureMiddlewareHashes = procedureUnits
		.map((unit) => unit.middlewares)
		.filter((unit) => unit.length > 0)
	const uniqueMiddlewareCombinations = new Set(procedureMiddlewareHashes)

	const middlewaresProcedureIdMap = new Map()
	for (const middlewares of uniqueMiddlewareCombinations.values()) {
		const middlewaresHash = middlewares.map((middleware) => middleware.hash).join('_')
		middlewaresProcedureIdMap.set(middlewares, `procedure_${middlewaresHash}`)
	}

	const routerShape: RouterShape = {
		tag: 'router',
		units: {},
		prefix: '',
	}

	for (const unit of procedureUnits) {
		const pathParts = unit.pathText.split('.')
		addProcedure(routerShape, unit, pathParts)
	}

	const childRouterUnits: RouterUnit[] = []
	const mergeRouterUnits: RouterUnit[] = []

	for (const unit of routerUnits) {
		if (unit.prefix) childRouterUnits.push(unit)
		else mergeRouterUnits.push(unit)
	}

	for (const unit of childRouterUnits) {
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		const pathParts = unit.prefix!.split('.').filter((value) => value !== '')
		addRouter(routerShape, unit, pathParts)
	}

	const middlewareUnits = units.filter((unit): unit is MiddlewareUnit => unit.tag === 'middleware')

	const ancestors = topNode.getAncestors()
	const topLevelNode = ancestors[ancestors.length - 2]

	let insertionIndex = topLevelNode.getChildIndex()
	for (const unit of middlewareUnits) {
		sourceFile.insertVariableStatement(insertionIndex, {
			declarationKind: VariableDeclarationKind.Const,
			declarations: [{
				name: unit.id,
				initializer: `t.middleware(${unit.body})`,
			}],
		}).formatText()
		insertionIndex += 1
	}

	for (const [middlewares, procedureId] of middlewaresProcedureIdMap.entries()) {
		const middlewareUses: string[] = []
		for (const middleware of middlewares) {
			middlewareUses.push(`.use(${middleware.id})`)
		}
		sourceFile.insertVariableStatement(insertionIndex, {
			declarationKind: VariableDeclarationKind.Const,
			declarations: [{
				name: procedureId,
				initializer: `${config.baseProcedure}${middlewareUses.join('')}`,
			}],
		})
		insertionIndex += 1
	}

	if (mergeRouterUnits.length > 0) {
		const writer = project.createWriter()
		writeShape({ writer, procedureOrRouter: { router: routerShape }, middlewaresProcedureIdMap, config })

		const newRouterText = writer.toString()
		const hash = getStringHash(newRouterText)
		const newRouterId = `router_${hash}`
		sourceFile.insertVariableStatement(insertionIndex, {
			declarationKind: VariableDeclarationKind.Const,
			declarations: [{
				name: newRouterId,
				initializer: newRouterText,
			}],
		})

		topNode.replaceWithText((writer) => {
			writer.write('t.mergeRouters(')
				.write(newRouterId)
				.write(',')
				.write(
					mergeRouterUnits.map((unit) => unit.identifier).join(', '),
				).write(')')
		})
	} else {
		topNode.replaceWithText((writer) => {
			writeShape({
				writer,
				procedureOrRouter: { router: routerShape },
				middlewaresProcedureIdMap,
				config,
			})
		})
	}
}
