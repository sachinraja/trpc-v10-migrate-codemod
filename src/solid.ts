import { CallExpression, Node, ObjectLiteralElementLike } from 'ts-morph'
import { getStringFromStringOrArrayLiteral } from './utils.js'

export const handleSolidHookCall = (
	type: string,
	callExpression: CallExpression,
) => {
	if (type === 'createQuery') {
		const arguments_ = callExpression.getArguments()
		const [pathArgument, configArgument] = arguments_
		if (Node.isArrowFunction(pathArgument)) {
			const path = getStringFromStringOrArrayLiteral(pathArgument.getBody())
			callExpression.removeArgument(pathArgument)
			return path
		}
		if (Node.isObjectLiteralExpression(configArgument)) {
			const trpcPropertyNames = ['context', 'ssr']

			const trpcProperties = trpcPropertyNames
				.map((property) => configArgument.getProperty(property))
				.filter((node): node is ObjectLiteralElementLike => !!node)

			configArgument.addPropertyAssignment({
				name: 'trpc',
				initializer: `{${trpcProperties.map((v) => v.getText()).join(',')}}`,
			})

			for (const property of trpcProperties) {
				property.remove()
			}

			configArgument.formatText()
		}

		if (Node.isArrayLiteralExpression(pathArgument)) {
			const elements = pathArgument.getElements()

			callExpression.insertArgument(0, elements[1].getText())
			const pathElement = elements[0]

			if (Node.isStringLiteral(pathElement)) {
				const path = pathElement.getLiteralText()
				callExpression.removeArgument(pathArgument)
				return path
			}
		}
	}

	if (type === 'createMutation') {
		const arguments_ = callExpression.getArguments()
		const pathArgument = arguments_[0]
		if (Node.isArrowFunction(pathArgument)) {
			const path = getStringFromStringOrArrayLiteral(pathArgument.getBody())
			callExpression.removeArgument(pathArgument)
			return path
		}
	}
}
