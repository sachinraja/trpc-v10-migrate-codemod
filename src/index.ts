import { Node, Project } from 'ts-morph'
import { handleCallerCall } from './caller.js'
import {
	contextHelpers,
	contextHelpersToRename,
	handleContextHelperCall,
	handleReactHookCall,
	OldContextHelper,
} from './react.js'
import { getRouterUnits, writeNewRouter } from './server.js'
import { MigrateConfig } from './types.js'

const resolveConfig = (config: Partial<MigrateConfig>): MigrateConfig => {
	return {
		reactNamespace: ['trpc'],
		callerNamespace: ['caller'],
		contextNamespace: ['utils'],
		routerFactory: ['router'],
		tsconfigPath: 'tsconfig.json',
		baseProcedure: 't.procedure',
		serverImports: [],
		removeServerImports: [],
		...config,
	}
}

export const transformv10Migration = async (config: Partial<MigrateConfig>) => {
	const resolvedConfig = resolveConfig(config)

	const project = new Project({
		tsConfigFilePath: resolvedConfig.tsconfigPath,
	})

	const sourceFiles = project.getSourceFiles()

	const allMigratedRouters: { filePath: string; identifier: string }[] = []

	await Promise.all(
		sourceFiles.map(async (sourceFile) => {
			const sourceFileMigratedRouters: string[] = []

			const filePath = sourceFile.getFilePath()
			let serverHasChanged = false

			sourceFile.forEachDescendant((node) => {
				if (!Node.isCallExpression(node)) return
				const firstChild = node.getFirstChild()

				if (Node.isIdentifier(firstChild)) {
					const firstChildText = firstChild.getText()
					if (resolvedConfig.reactNamespace.includes('') && ['useQuery', 'useMutation'].includes(firstChildText)) {
						const path = handleReactHookCall(firstChildText, node)
						if (!path) return

						firstChild.replaceWithText(`trpc.${path}.${firstChildText}`)
						return
					}

					if (resolvedConfig.routerFactory.includes(firstChildText)) {
						const { units, topNode } = getRouterUnits({ node })

						writeNewRouter({ project, units, sourceFile, topNode, config: resolvedConfig })
						const routerNameIdentifier = topNode.getParent()?.getFirstChild()
						if (Node.isIdentifier(routerNameIdentifier)) {
							const identifier = routerNameIdentifier.getText()
							sourceFileMigratedRouters.push(identifier)
							allMigratedRouters.push({ filePath, identifier })
						}
						serverHasChanged = true
						return
					}
				}

				if (!Node.isPropertyAccessExpression(firstChild)) return
				const callNamespaceOrCallExpression = firstChild.getFirstChild()

				if (Node.isIdentifier(callNamespaceOrCallExpression)) {
					const namespace = callNamespaceOrCallExpression.getText()

					const reactNamespaceMatch = resolvedConfig.reactNamespace.find((reactNamespace) =>
						reactNamespace === namespace
					)
					if (reactNamespaceMatch) {
						const procedureCallType = firstChild.getChildAtIndex(2).getText()

						const path = handleReactHookCall(procedureCallType, node)
						if (!path) return

						firstChild.replaceWithText(`${reactNamespaceMatch}.${path}.${procedureCallType}`)
						return
					}

					const callerNamespaceMatch = resolvedConfig.callerNamespace.find((callerNamespace) =>
						callerNamespace === namespace
					)
					if (callerNamespaceMatch) {
						const path = handleCallerCall(node)
						if (!path) return

						firstChild.replaceWithText(`${callerNamespaceMatch}.${path}`)
						return
					}

					if (resolvedConfig.contextNamespace.includes(namespace)) {
						const contextHelper = firstChild.getChildAtIndex(2).getText()
						if (!contextHelpers.includes(contextHelper as OldContextHelper)) return

						const renamedHelper = contextHelpersToRename[contextHelper as OldContextHelper]

						const { path, requiresUndefinedInput } = handleContextHelperCall(node)
						if (!path) return
						if (requiresUndefinedInput) {
							node.insertArgument(0, 'undefined')
						}

						firstChild.replaceWithText(`${resolvedConfig.contextNamespace}.${path}.${renamedHelper}`)
						return
					}
				}
			})

			if (serverHasChanged) {
				// add imports
				sourceFile.addImportDeclarations(resolvedConfig.serverImports)

				// remove imports
				for (const importDeclaration of sourceFile.getImportDeclarations()) {
					const removeNamedImports: string[] = []
					for (const removeImport of resolvedConfig.removeServerImports) {
						if (importDeclaration.getModuleSpecifierValue() === removeImport.moduleSpecifier) {
							removeNamedImports.push(...removeImport.namedImports)
						}
					}

					for (const namedImport of importDeclaration.getNamedImports()) {
						if (removeNamedImports.includes(namedImport.getName())) {
							namedImport.remove()
						}
					}

					if (importDeclaration.getNamedImports().length === 0) {
						importDeclaration.remove()
					}
				}
			}

			// always rename @trpc/react to @trpc/react-query
			for (const importDeclaration of sourceFile.getImportDeclarations()) {
				if (importDeclaration.getModuleSpecifierValue() === '@trpc/react') {
					importDeclaration.setModuleSpecifier('@trpc/react-query')
				}
			}

			console.log(`migrated ${filePath}`)
			for (const router of sourceFileMigratedRouters) {
				console.log(`  - migrated ${router}`)
			}
			await sourceFile.save()
		}),
	)

	console.log('\nMigration Summary:')
	for (const migratedRouter of allMigratedRouters) {
		console.log(`${migratedRouter.identifier} (${migratedRouter.filePath})`)
	}
}
