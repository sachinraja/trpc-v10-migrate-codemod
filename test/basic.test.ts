import { cp, readFile, rm } from 'node:fs/promises'
import path from 'node:path'
import { expect, it } from 'vitest'
import { transformv10Migration } from '../src'
import { importsMappingToImportDeclaration } from '../src/utils.js'

it('performs correct transformations', async () => {
	const snapshotDir = path.join(__dirname, 'snapshot')
	const tempDir = path.join(__dirname, 'temp')
	await cp(snapshotDir, tempDir, { recursive: true })

	await transformv10Migration({
		tsconfigPath: path.join(tempDir, 'tsconfig.test.json'),
		serverImports: importsMappingToImportDeclaration(['t:~/server/trpc', 'adminProcedure:~/server/trpc']),
		removeServerImports: importsMappingToImportDeclaration(['router:@trpc/server']),
		routerFactory: ['createProtectedRouter', 'router'],
		reactNamespace: ['trpc', ''],
	})

	const [transformedServer, transformedServer2, transformedReact] = await Promise.all([
		readFile(path.join(tempDir, 'server.ts'), 'utf8'),
		readFile(path.join(tempDir, 'server2.ts'), 'utf8'),
		readFile(path.join(tempDir, 'react.ts'), 'utf8'),
	])

	expect(transformedServer).toMatchSnapshot()
	expect(transformedServer2).toMatchSnapshot()
	expect(transformedReact).toMatchSnapshot()

	await rm(tempDir, { recursive: true })
})
