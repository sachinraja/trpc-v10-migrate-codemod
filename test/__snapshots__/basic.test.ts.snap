// Vitest Snapshot v1

exports[`performs correct transformations 1`] = `
"import { TRPCError } from '@trpc/server'
import { z } from 'zod'
import { exampleRouter } from './server2.js'
import { t } from \\"~/server/trpc\\";
import { adminProcedure } from \\"~/server/trpc\\";

const middleware_300161978 = t.middleware(async ({ ctx, next }) => {
    if (!ctx.user?.isAdmin) {
        throw new TRPCError({ code: 'UNAUTHORIZED' })
    }
    return next()
});
const procedure_300161978 = t.procedure.use(middleware_300161978);
const appRouter = t.router({
    hello: t.procedure.output(z.string()).query(() => {
        return 'world'
    }),
    post: t.procedure.query(() => {
        return 'post'
    }),post: t.procedure.mutation(() => {
        return 'post'
    }),post: t.router({
        create: t.procedure.input(z.object({ title: z.string() })).mutation(async ({ input }) => {
            return {
            	id: 1,
            	title: input.title,
            }
        }),
        byId: t.procedure.input(z.object({ id: z.string() })).query(({ input }) => {
        			return {
        				id: input.id,
        				title: 'hello',
        			}
        		}),
        getTitle: t.procedure.input(z.object({ title: z.string() })).query(({ input }) => {
        			return input.title
        		}),
    }),
    secretPlace: procedure_300161978.query(() => {
        return 'a key'
    }),
    example: t.router({
    }),
})


const caller = appRouter.createCaller({})
await caller.post.byId({ id: '1923071203' })
"
`;

exports[`performs correct transformations 2`] = `
"import { t } from \\"~/server/trpc\\";
import { adminProcedure } from \\"~/server/trpc\\";

const createProtectedRouter = router

export const exampleRouter = t.router({
    thing: t.procedure.query(() => {
        return 'example'
    }),
})

"
`;

exports[`performs correct transformations 3`] = `
"import { createTRPCReact } from '@trpc/react-query'
import { trpc } from '~/utils/trpc'

function Component() {
    const data = trpc.post.byId.useQuery({ id: 1 }, {
        onSuccess() {
            console.log('success')
        },
        trpc: {
            context: {
                foo: 'bar',
            }
        }
    })

	trpc.post.getTitle.useQuery({ title: 'hello' })

    trpc.hello.useQuery(undefined, {
        thing: true,
    })

    trpc.hello.useQuery(undefined, {
        thing: true,
        trpc: {
            ssr: true
        }
    })

	const mutation = trpc.post.create.useMutation({
		onSuccess() {
			console.log('success')
		},
	})

	const utils = trpc.useContext()

	utils.post.byId.invalidate(undefined, { active: true })
	utils.post.byId.cancel()
	utils.prefetchQuery([path])
	utils.post.byId.prefetch({ id: 1 })
	utils.post.getTitle.prefetch({ title: 'hello' })
}
"
`;
